### ISSUE TITLE
Unable to use combinators on prototypes.

### ISSUE THREAD

**Issue by willseitz** [Microsoft/TypeScript#3933]
I can not apply combinators to prototype. A combinator is a function that takes a function and returns another function. I use these to divide up functionality. I can use them in TypeScript, but they do not go on the prototype. I struggled to get them on the prototype, but TypeScript always throws an error. Here is an example of what I was trying to do. Inspired by TypeScript unions types I found many cases where I would like to pass T or T[]. I needed to write the code to handle this many places. So I wrote a combinator to handle this. So I would have:
```javascript
functionThatCanHandleBoth = makeFunctionToHandleArrayAndSingle(handleSingleFunction)
```
I want to put this on the prototype because I want to use these functions in objects that have many instances.

**Comment by mhegazy** [comment]
If i understand the issue correctly, you want to consolidate handling of parameter in one place, how about using decorators,

```ts
class C {
    @makeFunctionToHandleArrayAndSingle
     handleSingleFunction<T>(a:T) {

    }
}
```

also you can put it directelly on the prototype if you want to:

```ts
class C {
    handler: <T>(a: T) => void;
}

C.prototype.handler = makeFunctionToHandleArrayAndSingle(handleSingleFunction);
```

**Comment by willseitz** [comment]
Thanks for the response. Both those approaches work (I had tried the second, but decorators were new to me), but I couldn't get either to preserve context aka 'this' at the time. I discovered that this issue was actually in how I was making the decorator and not in the transpiled code. For reference here is a sample:

```javascript```
var decorate = function(func) {
    console.log("decoration being set up");
    return function() {
        console.log("/-----\\");
        var retval = func.apply(this,arguments);
        console.log("\\-----/");
        return retval;
    }
}

class Dog {
	constructor(private barkNoise: string) {
		
	}
	bark() {
		console.log(this.barkNoise);
	}
}


Dog.prototype.bark = decorate(Dog.prototype.bark);

var littleDog = new Dog("Yip");
var bigDog = new Dog("Woof");

littleDog.bark();
littleDog.bark();
littleDog.bark();
bigDog.bark();
```
